core/src/main/java/org/lflang/generator/docker/DockerGenerator.java

```diff
@@ -24,31 +32,174 @@ public DockerGenerator(LFGeneratorContext context) {
-  /** Return the default base image. */
-  public abstract String defaultImage();
+  /** Return the default compile commands for the C docker container. */
+  protected abstract List<String> defaultBuildCommands();
 
-  /** Return the selected base image, or the default one if none was selected. */
-  public String baseImage() {
-    var baseImage = context.getTargetConfig().get(DockerProperty.INSTANCE).from();
-    if (baseImage != null && !baseImage.isEmpty()) {
-      return baseImage;
+  /** Return the commands used to build */
+  protected List<String> getBuildCommands() {
+    if (context.getTargetConfig().isSupported(BuildCommandsProperty.INSTANCE)) {
+      var customBuildCommands = context.getTargetConfig().get(BuildCommandsProperty.INSTANCE);
+      if (customBuildCommands != null && !customBuildCommands.isEmpty()) {
+        return customBuildCommands;
+      }
     }
-    return defaultImage();
+    return defaultBuildCommands();
+  }
+
+  /** Return the command that sources the pre-build script, if there is one. */
+  protected List<String> getPreBuildCommand() {
+    var script = context.getTargetConfig().get(DockerProperty.INSTANCE).preBuildScript();
+    if (!script.isEmpty()) {
+      return List.of("source src-gen/" + script);
+    }
+    return List.of();
+  }
+
+  /** Return the command that sources the post-build script, if there is one. */
+  protected List<String> getPostBuildCommand() {
+    var script = context.getTargetConfig().get(DockerProperty.INSTANCE).postBuildScript();
+    if (!script.isEmpty()) {
+      return List.of("source src-gen/" + script);
+    }
+    return List.of();
+  }
+
+  /** Generate a header to print at the top of the Dockerfile. */
+  protected String generateHeader() {
+    return """
+           # Generated by the Lingua Franca compiler version %s
+           #   - Docs: https://www.lf-lang.org/docs/handbook/containerized-execution"
+           """
+        .formatted(LocalStrings.VERSION);
+  }
+
+  /** Return the Docker RUN command used for building. */
+  protected String generateRunForBuild() {
+    return "RUN "
+        + StringUtil.joinObjects(
+            Stream.of(
+                    List.of("set -ex"),
+                    getPreBuildCommand(),
+                    getBuildCommands(),
+                    getPostBuildCommand())
+                .flatMap(java.util.Collection::stream)
+                .collect(Collectors.toList()),
+            " \\\n\t&& ");
+  }
+
+  /** Return the ENTRYPOINT command. */
+  protected String generateEntryPoint() {
+    return "ENTRYPOINT ["
+        + getEntryPointCommands().stream()
+            .map(cmd -> "\"" + cmd + "\"")
+            .collect(Collectors.joining(","))
+        + "]";
+  }
+
+  /** Return a COPY command to copy the executable from the builder to the runner. */
+  protected String generateCopyOfExecutable() {
+    var lfModuleName = context.getFileConfig().name;
+    return "COPY --from=builder /lingua-franca/%s/bin/%s ./bin/%s"
```

Comment:
```suggestion
    return "COPY --from=builder /lingua-franca/%s/bin/%s ./bin/%s"  // safe because context.getFileConfig().name never contains spaces
```

Alternatively, we could just wrap the argument in quotes.